federatedscope/llm/model/adapter_builder.py

```diff
@@ -0,0 +1,181 @@
+
+            config = LoRAConfig(r=8, alpha=16)
+            model.add_adapter("lora_adapter", config=config)
+            model.train_adapter(['lora_adapter'])
+        elif adapter == 'bottleneck':
+            from transformers.adapters import AdapterConfig
+
+            config = AdapterConfig(mh_adapter=True,
+                                   output_adapter=True,
+                                   reduction_factor=16,
+                                   non_linearity="relu")
+            model.add_adapter("bottleneck_adapter", config=config)
+            model.train_adapter(['bottleneck_adapter'])
+        elif adapter == 'lang':
+            from transformers.adapters import PfeifferInvConfig
+
+            config = PfeifferInvConfig()
+            model.add_adapter("lang_adapter", config=config)
+            model.train_adapter(['lang_adapter'])
+        elif adapter == 'prefix':
+            from transformers.adapters import PrefixTuningConfig
+
+            config = PrefixTuningConfig(flat=False, prefix_length=30)
+            model.add_adapter("prefix_tuning", config=config)
+            model.train_adapter(['prefix_tuning'])
+        elif adapter == 'compacter':
+            from transformers.adapters import CompacterConfig
+
+            config = CompacterConfig()
+            model.add_adapter("dummy", config=config)
+            model.train_adapter(['dummy'])
+        elif adapter == 'ia_3':
+            from transformers.adapters import IA3Config
+
+            config = IA3Config()
+            model.add_adapter("ia3_adapter", config=config)
+            model.train_adapter(['ia3_adapter'])
+        elif adapter == 'union':
+            from transformers.adapters import AdapterConfig, ConfigUnion
+
+            # TODO: configure these args in cfg
+            config = ConfigUnion(
+                AdapterConfig(mh_adapter=True,
+                              output_adapter=False,
+                              reduction_factor=16,
+                              non_linearity="relu"),
+                AdapterConfig(mh_adapter=False,
+                              output_adapter=True,
+                              reduction_factor=2,
+                              non_linearity="relu"),
+            )
+            model.add_adapter("union_adapter", config=config)
+            model.train_adapter(['union_adapter'])
+        elif adapter == 'mam':
+            from transformers.adapters import \
+                ConfigUnion, ParallelConfig, PrefixTuningConfig
+
+            config = ConfigUnion(
+                PrefixTuningConfig(bottleneck_size=800),
+                ParallelConfig(),
+            )
+            model.add_adapter("mam_adapter", config=config)
+            model.train_adapter(['mam_adapter'])
+        else:
+            raise NameError(
+                f"There is no adapter named {adapter} in {package}")
+    else:
+        raise NotImplementedError
+    return model
+
+
+class AdapterModel(nn.Module):
```

Comment:
Same concern as weirui for unnecessary GPU [usage](https://github.com/alibaba/FederatedScope/pull/607#issuecomment-1556879934). We need to implement a overridden `to` function that only move the trainable part

Reply:
Here we use model.half() instead.

Reply:
And since we do not use deepcopy for the `base_model`, thus there is no need to override `to` function